RFC-0005-structured-kernel-definitions.md

```diff
@@ -0,0 +1,323 @@
+# Structured kernel definitions in PyTorch
+
+This is a proposal for a new code generation facility for writing
+kernels in PyTorch, where we will automatically generate
+easy-to-get-wrong boilerplate for functional (`add`), inplace (`add_`) and
+out (`add_out`) variants of functions, as well as common code (device
+guards, version counter tracking). The net result is you only need to
+write a shape checking function and an out-kernel when writing a
+function.
+
+**NOW HAS A PROTOTYPE** https://github.com/pytorch/pytorch/pull/45277 
+
+## Goals
+
+* **Be an opt-in mechanism**: always OK to continue to write code as it
+  is today. This ensures that third-party backend extenders also can
+  make things work without making use of codegen
+
+* **Reduce shape checking boilerplate**: make it easier to share common
+  shape checking code between CPU and CUDA implementations, as well as
+  with out-of-tree backend implementations
+
+* **Reduce functional/inplace/out boilerplate**: avoid having to write
+  foo, foo_, foo_out variants for every function; make it harder to
+  forget to add foo_out variant when it is appropriate
+
+* **Be able to run forward shape computation without kernel**: so
+  compilers and static runtimes can easily compute sizes of all elements
+  in the framework without having to actually run the computation
+
+* **Provide entry point for static runtime**: provide public API for
+  accessing operators directly, bypassing output allocation, shape
+  checking, device guards.
+
+* **Bypass dispatcher overhead**: in high performance cases, e.g., core
+  framework operators, reduce the number of redispatches to improve
+  performance
+
+* **Unify with version counter bumps and view metadata tracking**: these
+  are currently done in autograd generated code but must be performed
+  unconditionally even when autograd is disabled. This logic to be
+  incorporated with logic here
+
+* **Work with mobile selective build**: mobile selective build is
+  implemented in codegen, and implementation strategy must be compatible
+  with design constraints in mobile
+
+## Non-goals
+
+* **Directly handle TensorIterator**: TensorIterator is a uniquely
+  complicated mechanism, and we think it should be designed for on its
+  own.  However, we think that this is an important case and should
+  be handled sooner rather than later.
+
+* **Be zero runtime cost all the time**: We are willing to give up some
+  runtime efficiency for cleaner code. We are willing to give up
+  efficiency *by default* for out of tree implementers, unless they opt
+  in to higher performance. There is always an escape hatch to be high
+  performance if absolutely necessary
+
+* **No codegen**: As long as it is possible to implement things out of
```

Comment:
As long as we use this mechanism internally for all ops, we have meta functions for them defined and backends adding new kernels don't need to care about it. Only backends that add new operators would have to use it. Did I understand this correctly?

Reply:
That's right. So if you don't care about performance, you don't even have to lift a finger.

Reply:
wait I think I misunderstood this. To get device guards, you'll still have to go through the dispatch key for those backends, but not if calling into internal backends like CPU. How do you selectively make that dispatch key fallthrough for some backends but not for others?

Reply:
I claimed this would work in our meeting, but rereading your comment, I take it back, you really do need a per backend key here :(